projects/lexical_shared/lib/lexical/document/store.ex

```diff
@@ -1,97 +1,148 @@
+
+  @type start_opts :: [start_opt]
+  @type start_opt :: {:derive, derivations}
+
   defmodule State do
     @moduledoc false
+
     alias Lexical.Document
+    alias Lexical.Document.Store
+
     require Logger
 
-    defstruct temporary_open_refs: %{}
+    import Record
+
+    defstruct open: %{}, temporary_open_refs: %{}, derivation_funs: %{}
+
     @type t :: %__MODULE__{}
 
-    @table_name Document.Store
+    defrecord :open_doc, document: nil, derived: %{}
 
-    def new do
-      :ets.new(@table_name, [:named_table, :set, :protected, read_concurrency: true])
+    def new(opts \\ []) do
+      [derive: funs] =
+        opts
+        |> Keyword.validate!(derive: [])
+        |> Keyword.take([:derive])
 
-      %__MODULE__{}
+      %__MODULE__{derivation_funs: Map.new(funs)}
     end
 
-    @spec fetch(Lexical.uri()) :: {:ok, Document.t()} | {:error, :not_open}
-    def fetch(uri) do
-      case ets_fetch(uri, :any) do
-        {:ok, _} = success -> success
-        :error -> {:error, :not_open}
+    @spec fetch(t, Lexical.uri()) :: {:ok, Document.t(), t} | {:error, :not_open}
+    def fetch(%__MODULE__{} = store, uri) do
+      case store.open do
+        %{^uri => open_doc(document: document)} -> {:ok, document, store}
+        _ -> {:error, :not_open}
+      end
+    end
+
+    @spec fetch(t, Lexical.uri(), Store.derivation_key()) ::
+            {:ok, Document.t(), Store.derived_value(), t} | {:error, :not_open}
+    def fetch(%__MODULE__{} = store, uri, key) do
+      case store.open do
+        %{^uri => open_doc(document: document, derived: %{^key => derivation})} ->
+          {:ok, document, derivation, store}
+
+        %{^uri => open_doc(document: document, derived: derived)} ->
+          derivation = derive(store, key, document)
+          derived = Map.put(derived, key, derivation)
+          store = put_open_doc(store, document, derived)
+          {:ok, document, derivation, store}
+
+        _ ->
+          {:error, :not_open}
       end
     end
 
-    @spec save(t, Lexical.uri()) :: {:ok, t()} | {:error, :not_open}
+    @spec save(t, Lexical.uri()) :: {:ok, t} | {:error, :not_open}
     def save(%__MODULE__{} = store, uri) do
-      case ets_fetch(uri, :sources) do
-        {:ok, document} ->
+      case store.open do
+        %{^uri => open_doc(document: document, derived: derived)} ->
           document = Document.mark_clean(document)
-          ets_put(uri, :sources, document)
+          store = put_open_doc(store, document, derived)
           {:ok, store}
 
-        :error ->
+        _ ->
           {:error, :not_open}
       end
     end
 
     @spec open(t, Lexical.uri(), String.t(), pos_integer()) :: {:ok, t} | {:error, :already_open}
+    def open(%__MODULE__{temporary_open_refs: refs} = store, uri, text, version)
+        when is_map_key(refs, uri) do
+      {_, store} =
+        store
+        |> maybe_cancel_ref(uri)
+        |> pop_open_doc(uri)
+
+      open(store, uri, text, version)
```

Comment:
this seems like it's doing a bit more work that is necessary, why not just cancel the ref?

Reply:
My thought was that this means you can open a temporarily open doc as permanent using different text/version. Perhaps I misunderstood what the desired behavior is here, though?

Reply:
hmm... opening things temporarily is done by the server itself, I didn't consider differing (possibly lower) versions. What you have is correct.